src/shims/sync.rs

```diff
@@ -550,10 +551,13 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             Ok(0)
         } else if writers != 0 {
             let reader_blockset = rwlock_get_or_create_reader_blockset(this, rwlock_op)?;
-            rwlock_set_writers(this, rwlock_op, Scalar::from_u32(0))?;
+            // We are prioritizing writers here against the readers. As a
+            // result, not only readers can starve writers, but also writers can
+            // starve readers.
             if let Some(_writer) = this.unblock_some_thread(writer_blockset)? {
                 rwlock_set_writers(this, rwlock_op, Scalar::from_u32(1))?;
```

Comment:
Can't this be `assert!(writers == 1)` or so? You don't do a write when a mutex is handed from one thread to another; there is no reason to do a write here, either.